src/supa/documents/topology.py

```diff
@@ -0,0 +1,153 @@
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Generate a NSI topology document.
+
+Example topology document:
+
+TODO include example topology document
+"""
+from datetime import timedelta
+from typing import Union
+
+import cherrypy
+import structlog
+from lxml.etree import Element, QName, SubElement, tostring  # noqa: S410
+
+from supa import settings
+from supa.db.model import Topology
+from supa.nrm.backend import get_topology
+from supa.util.timestamp import current_timestamp
+
+log = structlog.get_logger(__name__)
+
+
+def refresh_topology() -> None:
+    """Refresh list of STP's in the database with topology from NRM, skip if configured for manual topology."""
+    if settings.manual_topology:
+        log.debug("skipping topology refresh", manual_topology=settings.manual_topology)
+        return
+
+    from supa.db.session import db_session
+
+    nrm_stps = get_topology()
+    nrm_stp_ids = [nrm_stp.stp_id for nrm_stp in nrm_stps]
+    with db_session() as session:
+        for nrm_stp in nrm_stps:
+            if nrm_stp.topology != settings.topology:
+                log.debug("skip STP with unknown topology", stp=nrm_stp.stp_id, topology=nrm_stp.topology)
+            else:
+                stp = session.query(Topology).filter(Topology.stp_id == nrm_stp.stp_id).one_or_none()
+                if stp:
+                    log.debug("update existing STP", stp=nrm_stp.stp_id)
+                    stp.port_id = nrm_stp.port_id
+                    stp.vlans = nrm_stp.vlans
+                    stp.description = nrm_stp.description
+                    stp.is_alias_in = nrm_stp.is_alias_in
+                    stp.is_alias_out = nrm_stp.is_alias_out
+                    stp.bandwidth = nrm_stp.bandwidth
+                    stp.enabled = nrm_stp.expose_in_topology
+                else:
+                    log.info("add new STP", stp=nrm_stp.stp_id)
+                    session.add(
+                        Topology(
+                            stp_id=nrm_stp.stp_id,
+                            port_id=nrm_stp.port_id,
+                            vlans=nrm_stp.vlans,
+                            description=nrm_stp.description,
+                            is_alias_in=nrm_stp.is_alias_in,
+                            is_alias_out=nrm_stp.is_alias_out,
+                            bandwidth=nrm_stp.bandwidth,
+                            enabled=nrm_stp.expose_in_topology,
+                        )
+                    )
+        for stp in session.query(Topology).filter(Topology.enabled):
+            if stp.stp_id not in nrm_stp_ids:
+                log.info("disable vanished STP", stp=stp.stp_id)
+                stp.enabled = False
+
```

Comment:
The save is probably done implicitly by sqlalchemy's autoflush but it's better to have an explicit `stp.save()` in here somewhere

Reply:
That is correct, the moment the database session is going out of scope the changes are committed to the database, or rolled back on error, see: https://github.com/workfloworchestrator/SuPA/blob/c82a1304fddaee707d710bafb43f75b55dae87af/src/supa/db/session.py#L130.
This pattern is used everywhere in the code, I do not see a reason to add an explicit save() here.